### PR TITLE
introduce inline content for configs

### PR DESCRIPTION
**Pull request by ndeloof**
**What this PR does / why we need it**:
Introduce support for inlined content when defining a configuration file

**Which issue(s) this PR fixes**:
Fixes https://github.com/compose-spec/compose-spec/issues/346

### CODE REVIEW

File: 08-configs.md [hunk 1]
```diff
@@ -13,6 +13,8 @@ By default, the config:
 The top-level `configs` declaration defines or references configuration data that is granted to services in your Compose application. The source of the config is either `file` or `external`.
 
 - `file`: The config is created with the contents of the file at the specified path.
+- `environment`: The config is created with the value of an environment variable.
```
Comment (glours) [inline 08-configs.md:16]:
```suggestion
- `environment`: The config content is created with the value of an environment variable.
```

**Review by glours — APPROVED** [review]
Just as nit to make things clearer

**Comment by Bilge** [comment]
How does one benefit from this change? With the latest Docker Desktop 4.25.2 (129061), I get `configs.fpm Additional property content is not allowed`.

**Comment by polarathene** [comment]
> `configs.fpm Additional property content is not allowed`.

I also tried this since it's published on the docs (_[`configs` top-level](https://docs.docker.com/compose/compose-file/08-configs/) page, service [`configs` attribute](https://docs.docker.com/compose/compose-file/05-services/#configs)_) and there is no min version description, so I assumed it was already available at some point.

I looked through the [release notes](https://docs.docker.com/compose/release-notes) and **found nothing** (_not too fun with fairly generic keywords `configs` / `content`_).

Then I tried to track down the source from the docs:
- There is no direct link for contributions on the rendered docs page, but it takes you to `docker/docs` repo.
- While some of the compose docs are there, this particular one was missing. Searched the repo and found [hugo config ref](https://github.com/docker/docs/blob/c47c4a4ba18a686a3ac2b353c2aec452aa55bbaa/hugo.yaml#L282C13-L299).
- Now at `compose-spec/compose-spec`, found the [docs page content and `git blame`](https://github.com/compose-spec/compose-spec/blame/ca5c01413d11fed0241f7726e03f6d4540ef3bb7/08-configs.md#L17) to find out it was merged in Nov 7th here.
  - `compose-go 1.20.1` (Nov 14th) [released with the feature](https://github.com/compose-spec/compose-go/releases/tag/v1.20.1).
  - [`docker/compose` upgraded to that dep](https://github.com/docker/compose/pull/11187) on the same day. Then [released v2.23.1](https://github.com/docker/compose/releases/tag/v2.23.1) on Nov 16th.
    - The release notes don't mention the feature, nor does the [Docker Compose docs release notes page](https://docs.docker.com/compose/release-notes/#2231).

That was a bit of a mission 😅 Didn't realize the feature was that new. I'm on v2.21.0 (Aug 30th).

I was on 4.23.0 (120376), but now also upgraded to the latest version is 4.25.2 (129061), but unfortunately that's not new enough 😆 

---

# TL;DR

> How does one benefit from this change?

[Docker Desktop 4.25.0](https://docs.docker.com/desktop/release-notes/#4250) (released Oct 26th) provides Docker Compose 2.23.0 (`docker compose version`).

While 4.25.2 (Nov 21st) probably could have bumped the compose version, we'll probably have to **wait a little bit longer** for a new release to handle that.

**Comment by selivan** [comment]
@ndeloof 

I would like to suggest upgrading compose specification version for changes like this one.

Just now I've ran into this situation:
- I write `docker-compose.yml` according to official docs
- I specify the latest available `version: "3.8"`
- `docker compose` does not complain about incompatibility with this version
- And yet I get an error `Additional property content is not allowed`

That is confusing.

**Comment by ndeloof** [comment]
> I specify the latest available version: "3.8"

`version` is obsolete with compose-spec, it is ignored. Compose spec isn't versioned _by design_. The compose (tool) version you're using will let you know it can't parse your compose file if you use a recent addition.

> docker compose does not complain about incompatibility with this version
And yet I get an error Additional property content is not allowed

This is exactly what "complain about incompatibility" is about. Maybe we should add "_please upgrade docker compose_" to this message.

**Comment by selivan** [comment]
@ndeloof 

Unfortunately, there is no way to distinguish between _"please upgrade docker compose"_ and _"you mistyped a keyword or used it on incorrect nesting level"_.

> `version` is obsolete with compose-spec, it is ignored.

That is sad 😢 

Look at terraform and terraform providers, for example. Each version explicitly has it's own specification. As long as you follow that specification for selected version - you will not run into compatibility issues.
